aptos-move/framework/src/natives/cryptography/algebra/arithmetics/add.rs

```diff
@@ -0,0 +1,73 @@
+// Copyright © Aptos Foundation
+
+use crate::{
+    abort_unless_arithmetics_enabled_for_structure, abort_unless_feature_flag_enabled,
+    ark_binary_op_internal,
+    natives::{
+        cryptography::algebra::{
+            abort_invariant_violated, feature_flag_from_structure, gas::GasParameters,
+            AlgebraContext, Structure, MOVE_ABORT_CODE_NOT_IMPLEMENTED,
+        },
+        helpers::{SafeNativeContext, SafeNativeError, SafeNativeResult},
+    },
+    safe_borrow_element, safely_pop_arg, store_element, structure_from_ty_arg,
+};
+use move_core_types::gas_algebra::NumArgs;
+use move_vm_types::{loaded_data::runtime_types::Type, values::Value};
+use smallvec::{smallvec, SmallVec};
+use std::{
+    collections::VecDeque,
+    ops::{Add, Mul},
+    rc::Rc,
+};
+
+pub fn add_internal(
+    gas_params: &GasParameters,
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    assert_eq!(1, ty_args.len());
+    let structure_opt = structure_from_ty_arg!(context, &ty_args[0]);
+    abort_unless_arithmetics_enabled_for_structure!(context, structure_opt);
+    match structure_opt {
```

Comment:
It's sad that this can't be done on the level of Move via a trait type system, which statically resolves to the right functions. Good example why we later may want this feature in Move.